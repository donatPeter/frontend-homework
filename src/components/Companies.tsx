import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { ReduxState, Company } from "../types/types";
import { getCompanies, getSelectedCompanyId } from "../store/selectors";
import { CompanyLink } from "./CompanyLink";
import styled from "styled-components";
import {
  setSelectedCompanyId,
  toggleDropdownMenuVisibility,
} from "../store/actions";

export const Title = styled.p`
  text-transform: uppercase;
  color: #717175;
  margin: 12px 16px 4px 16px;
  font-weight: bold;
`;

const CompaniesContainer = styled.div`
  width: 100%;
  border-bottom: 1px solid #dcdce0;
`;

type ReduxProps = {
  companies: Array<Company>;
  selectedCompanyId: number | null;
};

type DispatchProps = {
  setSelectedCompanyId: (id: number) => void;
  toggleDropdownMenuVisibility: () => void;
};

export const Companies = ({
  companies,
  selectedCompanyId,
  setSelectedCompanyId,
  toggleDropdownMenuVisibility,
}: ReduxProps & DispatchProps) => {
  const handleOnClick = (companyId: number) => {
    setSelectedCompanyId(companyId);
    if (selectedCompanyId !== companyId) {
      toggleDropdownMenuVisibility();
    }
  };
  return (
    <CompaniesContainer>
      <Title>Your companies</Title>

      {companies.map((company) => (
        <CompanyLink
          key={company.id}
          company={company}
          isSelected={selectedCompanyId === company.id}
          onClick={handleOnClick}
        />
      ))}
    </CompaniesContainer>
  );
};

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    companies: getCompanies,
    selectedCompanyId: getSelectedCompanyId,
  }),
  { setSelectedCompanyId, toggleDropdownMenuVisibility }
)(Companies);
