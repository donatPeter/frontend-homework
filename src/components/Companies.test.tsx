import React from "react";
import { shallow } from "enzyme";
import { Companies, Title } from "./Companies";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("<Companies />", () => {
  const closeDropdownMock = jest.fn();
  const setSelectedCompanyIdMock = jest.fn();
  const companies = [
    { id: 1, name: "Dummy company" },
    { id: 2, name: "Smarty company" },
  ];

  it("renders list of company links", () => {
    expect(
      shallow(
        <Companies
          setSelectedCompanyId={setSelectedCompanyIdMock}
          closeDropdown={closeDropdownMock}
          selectedCompanyId={1}
          companies={companies}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders the title of the company list", () => {
    const wrapper = shallow(
      <Companies
        setSelectedCompanyId={setSelectedCompanyIdMock}
        closeDropdown={closeDropdownMock}
        selectedCompanyId={1}
        companies={companies}
      />
    );

    expect(wrapper.find(Title).text()).toBe("Your companies");
  });

  it("invoke setSelectedCompanyId and closeDropdown on click", () => {
    const { getByText } = render(
      <Companies
        setSelectedCompanyId={setSelectedCompanyIdMock}
        closeDropdown={closeDropdownMock}
        selectedCompanyId={1}
        companies={companies}
      />
    );

    expect(setSelectedCompanyIdMock).not.toHaveBeenCalled();
    expect(closeDropdownMock).not.toHaveBeenCalled();
    userEvent.click(getByText("Smarty company"));
    expect(setSelectedCompanyIdMock).toHaveBeenCalledWith(2);
    expect(closeDropdownMock).toHaveBeenCalledTimes(1);
  });
});
