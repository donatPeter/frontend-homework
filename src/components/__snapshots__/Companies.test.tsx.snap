// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Companies /> renders list of company links 1`] = `
<styled.div>
  <styled.p>
    Your companies
  </styled.p>
  <CompanyLink
    company={
      Object {
        "id": 1,
        "name": "Dummy company",
      }
    }
    isSelected={true}
    key="1"
    onClick={[Function]}
  />
  <CompanyLink
    company={
      Object {
        "id": 2,
        "name": "Smarty company",
      }
    }
    isSelected={false}
    key="2"
    onClick={[Function]}
  />
</styled.div>
`;
